torch/csrc/autograd/forward_grad.cpp

```diff
@@ -0,0 +1,76 @@
+#include <torch/csrc/autograd/forward_grad.h>
+
+namespace torch { namespace autograd {
+
+namespace {
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    static std::mutex all_forward_levels_mutex_;
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    static uint64_t next_forward_idx_ = 0;
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    static std::vector<std::shared_ptr<ForwardADLevel>> all_forward_levels_;
```

Comment:
What happens if two distinct threads are using forward AD, and not talking to each other? A global here seems like it would cause contention.

Reply:
The contention only happens at start and end of forward AD section, and it means that you can safely transmit dual tensors across threads (however!!! the relative level ordering is undefined)

Reply:
This design might be right. It has a lot of good properties; for example, if you start a forward AD in one thread, transmit some tensors to another worker thread, and then start a nested AD in that thread, the implicit synchronization on all starts will do the right thing (because the out-of-band synchronization from tensor transmission guarantees the relative ordering). But you can't assume that the forward AD levels will exit in an orderly fashion; you basically have implemented non-lexical forward AD.

Another possible design to roll around on your tongue is associate levels with a thread index. Levels from different threads are incomparable and forward AD will complain if you get a tensor in another thread that is associated with a context. Then there would need to be some way of transmitting a forward AD context to another thread so you could send forward AD tensors to another worker thread and work on it. (Of course, none of this would get implemented at first, but it would influence the design here.)